### DISCUSSION TITLE
[Question] Question about getting multi rigid contact data to simulate robot skin

### DISCUSSION THREAD

**Discussion by ZhangYi1999** [isaac-sim/IsaacLab#1630] [Q&A]
Hello, everyone. I am currently investigate how to simulate a robot skin module in isaac lab. The robot skin module is consist of multi small contact sensor cell and connected with each other to become a grid or hive shape. Like this,

![image](https://github.com/user-attachments/assets/d428f699-3ffe-487e-b897-9380cf151317)

In this case, each contact cell will have its own force sensor data. According to my understand about isaac sim and isaac lab, I think there are following possible way to simulate it properly:

## RigidContactView

Create a `omni.isaac.core.prims.RigidBody` for every cell, connected them with `UsdPhysics.FixedJoint`, and use a `omni.isaac.core.prims.RigidContactView` to query all cell in all environment at the same time. This is the same idea as `ContactSensor` from `omni.isaac.lab`.

However, this method doesn't work because when the amount is getting higher, it gives this error message.

```
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.plugin] PhysX error: PhysX Internal CUDA error. Simulation cannot continue! Error code 700!
, FILE /builds/omniverse/physics/physx/source/physx/src/NpScene.cpp, LINE 3059
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.plugin] Cuda context manager error, simulation will be stopped and new cuda context manager will be created.
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.tensors.plugin] CUDA error: an illegal memory access was encountered: ../../../extensions/runtime/source/omni.physx.tensors/plugins/gpu/GpuSimulationData.cpp: 152
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.tensors.plugin] CUDA context validation failed
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.tensors.plugin] CUDA error: an illegal memory access was encountered: ../../../extensions/runtime/source/omni.physx.tensors/plugins/gpu/GpuSimulationData.cpp: 199
2024-08-22 12:25:15 [25,841ms] [Error] [omni.physx.tensors.plugin] Failed to initialize GPU simulation data
Traceback (most recent call last):
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/assets/asset_base.py", line 92, in <lambda>
    lambda event, obj=weakref.proxy(self): obj._initialize_callback(event),
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/assets/asset_base.py", line 263, in _initialize_callback
    self._initialize_impl()
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/assets/articulation/articulation.py", line 946, in _initialize_impl
    self._physics_sim_view = physx.create_simulation_view(self._backend)
  File "/home/yi/.local/share/ov/pkg/isaac-sim-4.1.0/extsPhysics/omni.physics.tensors/omni/physics/tensors/impl/api.py", line 12, in create_simulation_view
    raise Exception("Failed to create simulation view backend")
Exception: Failed to create simulation view backend
Traceback (most recent call last):
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/sensor_base.py", line 64, in <lambda>
    lambda event, obj=weakref.proxy(self): obj._initialize_callback(event),
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/sensor_base.py", line 271, in _initialize_callback
    self._initialize_impl()
  File "/home/yi/Program/isaac_skin_simulation/exts/isaac_skin/isaac_skin/sensors/skin/skin.py", line 191, in _initialize_impl
    super()._initialize_impl()
  File "/home/yi/Program/IsaacLab/source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/sensor_base.py", line 225, in _initialize_impl
    self._is_outdated = torch.ones(self._num_envs, dtype=torch.bool, device=self._device)
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
2024-08-22 12:25:21 [31,781ms] [Error] [carb.graphics-vulkan.plugin] VkResult: ERROR_INITIALIZATION_FAILED
2024-08-22 12:25:21 [31,781ms] [Error] [carb.graphics-vulkan.plugin] vkCreateRayTracingPipelinesKHR failed.
2024-08-22 12:25:21 [31,781ms] [Error] [carb.graphics-vulkan.plugin] No Collection/Library can be nullptr with eCollection as compileType.
2024-08-22 12:25:21 [31,781ms] [Error] [gpu.foundation.plugin] Failed to create the raytracing pipeline.
```

I think it could be the issue of creating too many rigid body in an inefficient way.

## Callback function

Instead of creat a `RigidBody` for each cell, we can also create only one `RigidBody` for the whole group and consider each cell as an individual `Collider` under the overall rigid body.

Use a customed callback function, which can get the detailed collider information like the following code. This code comes from the example in `ContactReportDemo` and `ContactReportDirectAPIDemo` from `omni.physx.demos`

```python
    def update(self, stage, dt, viewport, physxIFace):
        if not self._contact_report_sub:
            if self.use_full_report:
                self._contact_report_sub = get_physx_simulation_interface().subscribe_full_contact_report_events(self._on_full_contact_report_event)
            else:
                self._contact_report_sub = get_physx_simulation_interface().subscribe_contact_report_events(self._on_contact_report_event)
        
    def _on_contact_report_event(self, contact_headers, contact_data):
        self.print_contacts(contact_headers, contact_data, None)
                        
    def _on_full_contact_report_event(self, contact_headers, contact_data, friction_anchors):
        self.print_contacts(contact_headers, contact_data, friction_anchors)

    def print_contacts(self, contact_headers, contact_data, friction_anchors):
        for contact_header in contact_headers:
            print("Got contact header type: " + str(contact_header.type))
            print("Actor0: " + str(PhysicsSchemaTools.intToSdfPath(contact_header.actor0)))
            print("Actor1: " + str(PhysicsSchemaTools.intToSdfPath(contact_header.actor1)))
            print("Collider0: " + str(PhysicsSchemaTools.intToSdfPath(contact_header.collider0)))
            print("Collider1: " + str(PhysicsSchemaTools.intToSdfPath(contact_header.collider1)))
            print("StageId: " + str(contact_header.stage_id))
            print("Number of contacts: " + str(contact_header.num_contact_data))
            
            contact_data_offset = contact_header.contact_data_offset
            num_contact_data = contact_header.num_contact_data
            
            for index in range(contact_data_offset, contact_data_offset + num_contact_data, 1):
                print("Contact:")
                print("Contact position: " + str(contact_data[index].position))
                print("Contact normal: " + str(contact_data[index].normal))
                print("Contact impulse: " + str(contact_data[index].impulse))
                print("Contact separation: " + str(contact_data[index].separation))
                print("Contact faceIndex0: " + str(contact_data[index].face_index0))
                print("Contact faceIndex1: " + str(contact_data[index].face_index1))
                print("Contact material0: " + str(PhysicsSchemaTools.intToSdfPath(contact_data[index].material0)))
                print("Contact material1: " + str(PhysicsSchemaTools.intToSdfPath(contact_data[index].material1)))

            if friction_anchors is not None:
                friction_anchors_data_offset = contact_header.friction_anchors_offset
                num_friction_anchor_data = contact_header.num_friction_anchors_data

                for index in range(friction_anchors_data_offset, friction_anchors_data_offset + num_friction_anchor_data, 1):
                    print("Friction anchor:")
                    print("Friction position: " + str(friction_anchors[index].position))
                    print("Friction impulse: " + str(friction_anchors[index].impulse))
```
However, this way doesn't make much sense since it should query the information one by one, which means if you have 10 env and each env has 100 cell, you need to run this call function 1000 times. What's more, I am not sure if this call function runs on cpu or gpu, and will it effect the preformance of RL or not.

## Calculate the cell assignment manually

Since `RigidContactView` can also report contact position, it will be possible to use the contact position to determine which cell is collided. However, this could increase the computation complexity.

## Question

My question is, is there a better way to do this or can we use any of the above idea but do it efficiently. Btw, is it possible that  the `get_contact_force_data` function in `RigidContactView` return collider information as well?
If anyone could give same advice, I will be super grateful. Thanks in advance!

**Comment by RandomOakForest** [comment]
Thanks for posting this. Our apologies we haven't been able to reply earlier. I'll move this to a _discussion_ for the team to follow up upon.

**Comment by YuHoChau** [comment]
I have the same question here. Any updates so far? Thanks.

**Comment by zhexulong** [comment]
Hi, I am considering expanding the contact sensor API. In addition to the existing aggregated data (such as net_forces, force_matrix, average contact points, etc.), a new per-contact raw output(provided by RigidContactView.get_contact_data()) will be added to meet the requirement of providing "for each contact point: position, normal force, and friction force".
Do you think this will be helpful to you?  @YuHoChau 
@ZhangYi1999 What you mentioned – "this method doesn't work because when the amount is getting higher, it gives this error message" – is probably because the "max_contact_data_count" was set too low when you called create_rigid_contact_view. This seems to be a bug.If you re-run it with CUDA_LAUNCH_BLOCKING=1, you may be able to see the specific location of the out-of-bounds behavior.
Additionally, I would like to ask @RandomOakForest : What level of accuracy do the raw outputs provided by PhysX RigidContactView possess, and what types of tasks can they reliably support? Additionally, are there any methods, tools, or metrics available to help us quantify or measure the accuracy of these raw outputs?

**Reply by YuHoChau** [reply]
Hi @zhexulong, apologize for my late reply. I agree it would be very helpful for the community for simulating such dense tactile arrays. Yet our concernes were the capbility of the simulator itself as we encountered the similar issue discussed here: https://github.com/linchangyi1/LocoTouch/issues/2. Do you have any idea that how to leverage the `contact sensor` in IsaacLab to achive this? I think if can directly use with this build-in function, it would be much easier than integrating other compliacted functions such as SDF (https://github.com/isaac-sim/IsaacGymEnvs/blob/tacsl/isaacgymenvs/tacsl_sensors/install/tacsl_setup.md, and https://github.com/NVlabs/vt-refine) Also for your reference: https://github.com/isaac-sim/IsaacLab/issues/3578#issuecomment-3383412544, yet so far I do not have enough time for solving it. But very happy to discuss and help, if you are interested in solving this together, please feel free to reach to me. Thanks.

**Reply by SarveshS325** [reply]
I have the same question. I have created small rigid bodies in the robot USD and attached them to links using fixed joints. This does work with the isaaclab contact sensor but slows down the simulation drastically. With 220 such tactile pad rigid bodies (and without actually instantiating the contact sensors), the simulation is a couple of orders of magnitude slower on a RTX A6000 Ada.

Has there been any progress on this? Any suggestion to make the performance better?

**Reply by YuHoChau** [reply]
Hi @SarveshS325 , you may try this method https://github.com/isaac-sim/IsaacLab/issues/3578#issuecomment-3383412544 but I never tried before... Would be greatly appreciated if you could share if this works or not. Meanwhile I recently achieved the SDF-based version of tactile array sensors following https://github.com/NVlabs/vt-refine, in IsaacLab. Would be happy to discuss.

<img width="1165" height="362" alt="image" src="https://github.com/user-attachments/assets/91b53d44-997c-41eb-a90e-6f77c581c712" />

**Reply by SarveshS325** [reply]
I did try the approach mentioned in the comment. I get an error like "No ContactReporter API found at the specified prim". This is expected because it seems that ContactReporter API requires both colliders and RigidBody applied to the prim and having a RigidBody as a child of another RigidBody is not supported. In short, that way will not work (the comment seems to be LLM generated to me)

I'll take a look at vt-refine and get back to you.  Appreciate your help

**Reply by SarveshS325** [reply]
vt-refine seems to be very similar to TacSL recently added to isaaclab_contrib

**Reply by YuHoChau** [reply]
Yes, SDF-based.

**Reply by SarveshS325** [reply]
> I have the same question. I have created small rigid bodies in the robot USD and attached them to links using fixed joints. This does work with the isaaclab contact sensor but slows down the simulation drastically. With 220 such tactile pad rigid bodies (and without actually instantiating the contact sensors), the simulation is a couple of orders of magnitude slower on a RTX A6000 Ada.
> 
> Has there been any progress on this? Any suggestion to make the performance better?

If anyone is trying the fixed joints approach and facing slow simulation rate, running the sim with `create_stage_in_memory=True` helps quite a lot.
